atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultSingleBucketSweepTask.java

```diff
@@ -0,0 +1,158 @@
+        }
+
+        BucketProgress existingBucketProgress =
+                bucketProgressStore.getBucketProgress(sweepableBucket.bucket()).orElse(BucketProgress.INITIAL_PROGRESS);
+
+        // This is inclusive.
+        long lastSweptTimestampInBucket =
+                sweepableBucket.timestampRange().startInclusive() + existingBucketProgress.timestampProgress();
+        if (isCompletelySwept(sweepableBucket.timestampRange().endExclusive(), lastSweptTimestampInBucket)) {
+            // The bucket is fully swept; it might still be returned here if we thought it was a candidate, or if
+            // the bucket state machine is still doing things
+            markBucketCompleteIfEligible(sweepableBucket);
+            return 0L;
+        }
+
+        if (sweepTimestampForIteration <= lastSweptTimestampInBucket) {
+            // The sweep timestamp has made progress in this partition, but we've swept everything up to it.
+            return 0L;
+        }
+
+        // TODO (jkong): Make use of the partial progress within a timestamp.
+        SweepBatchWithPartitionInfo sweepBatchWithPartitionInfo = sweepQueueReader.getNextBatchToSweep(
+                sweepableBucket.bucket().shardAndStrategy(),
+                lastSweptTimestampInBucket,
+                getEndOfSweepRange(sweepableBucket, sweepTimestampForIteration));
+        SweepBatch sweepBatch = sweepBatchWithPartitionInfo.sweepBatch();
+
+        ShardAndStrategy shardAndStrategy = sweepableBucket.bucket().shardAndStrategy();
+        sweepQueueDeleter.sweep(sweepBatch.writes(), Sweeper.of(shardAndStrategy));
+        targetedSweepMetrics.registerEntriesReadInBatch(shardAndStrategy, sweepBatch.entriesRead());
+
+        if (!sweepBatch.isEmpty()) {
+            log.debug(
+                    "Put {} ranged tombstones and swept up to timestamp {} for {}.",
+                    SafeArg.of("tombstones", sweepBatch.writes().size()),
+                    SafeArg.of("lastSweptTs", sweepBatch.lastSweptTimestamp()),
+                    SafeArg.of("shardStrategy", shardAndStrategy.toText()));
+        }
+
+        long lastTs = sweepBatch.lastSweptTimestamp();
+        sweepQueueCleaner.clean(
+                shardAndStrategy,
+                sweepBatchWithPartitionInfo.partitionsForPreviousLastSweptTsWithMinimumBound(
+                        lastSweptTimestampInBucket,
+                        sweepableBucket.timestampRange().startInclusive()),
+                lastTs,
+                sweepBatch.dedicatedRows());
+
+        long lastTsOffset = lastTs - sweepableBucket.timestampRange().startInclusive();
+
+        bucketProgressStore.updateBucketProgressToAtLeast(
+                sweepableBucket.bucket(), BucketProgress.createForTimestampProgress(lastTsOffset));
+        if (isCompletelySwept(sweepableBucket.timestampRange().endExclusive(), lastTs)) {
+            // we've finished the bucket!
+            markBucketCompleteIfEligible(sweepableBucket);
+        }
+
```

Comment:
note: SweepQueue increments the number of tombstones metric (L178). We don't do that here, is that intentional?

Reply:
Yep, that wasn't intentional - fixed. I think this was an artifact of a previous design where the task didn't do cleaning